File: GraphMatching/src/com/jakemadethis/graph/GraphFrame.java
package com.jakemadethis.graph;

import java.awt.BorderLayout;
import java.awt.CardLayout;
import java.awt.Color;
import java.awt.Component;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JPanel;

import com.jakemadethis.graph.Chooser.ChoiceListener;

import edu.uci.ics.jung.algorithms.layout.Layout;
import edu.uci.ics.jung.algorithms.layout.util.Relaxer;
import edu.uci.ics.jung.graph.Hypergraph;

public class GraphFrame extends JFrame implements ChoiceListener<Hypergraph<Vertex, Hyperedge>> {
	private JPanel sidebar;
	private JPanel cards;
	private JPanel menubar;
	private final Chooser<Hypergraph<Vertex, Hyperedge>> chooser;

	public GraphFrame(Chooser<Hypergraph<Vertex, Hyperedge>> chooser) {
		this.chooser = chooser;
		chooser.addChoiceListener(this);
		
		setLayout(new BorderLayout());
		sidebar = new JPanel();
		sidebar.setLayout(new BoxLayout(sidebar, BoxLayout.Y_AXIS));
		sidebar.setBackground(Color.WHITE);
		getContentPane().add(sidebar, BorderLayout.WEST);
		cards = new JPanel(new CardLayout());
		JPanel main = new JPanel();
		main.setLayout(new BoxLayout(main, BoxLayout.Y_AXIS));
		menubar = new JPanel();
		menubar.setBackground(Color.DARK_GRAY);
		menubar.add(new JButton("Test1"));
		menubar.add(new JButton("Test2"));
		main.add(menubar);
		main.add(cards);
		
		getContentPane().add(main, BorderLayout.CENTER);		
		setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);	
	}

	@Override
	public void makeChoice(Hypergraph<Vertex, Hyperedge> item) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void addChoice(Hypergraph<Vertex, Hyperedge> item) {
	}

	@Override
	public void removeChoice(Hypergraph<Vertex, Hyperedge> item) {
		// TODO Auto-generated method stub
		
	}
}

File: Graphite/src/com/jakemadethis/graphite/ui/GraphFrame.scala
package com.jakemadethis.graphite.ui

import javax.swing.JFrame
import edu.uci.ics.jung.graph.Hypergraph
import com.jakemadethis.graphite.graph._
import edu.uci.ics.jung.graph.Graph
import edu.uci.ics.jung.algorithms.layout.StaticLayout
import com.jakemadethis.graphite.visualization.AverageEdgeLayout
import edu.uci.ics.jung.algorithms.layout.util.RandomLocationTransformer
import java.awt.Dimension
import edu.uci.ics.jung.visualization.DefaultVisualizationModel
import javax.swing.WindowConstants
import scala.swing._
import com.jakemadethis.graphite.algorithm.{Derivation,HypergraphProduction}
import com.jakemadethis.graphite.algorithm.HypergraphGenerator
import java.awt.Color

class GraphFrame(devPaths : Seq[Derivation.Path[HypergraphProduction]]) extends MainFrame {
  
  type Model = DefaultVisualizationModel[Vertex, Hyperedge]
  val genGraphs = devPaths.map(new GeneratableGraph(_))
  
  val graphpanel = new GraphPanel(genGraphs.head.model)

  val graphButtons = new BoxPanel(Orientation.Vertical) {
    
    background = Color.WHITE
    
    var toggled : ToggleButton = null
    
    def setToggled(button : ToggleButton) {
      if (toggled != null) toggled.selected = false
      toggled = button
      toggled.selected = true
    }
    def clear() {
      toggled = null
      contents.clear
    }
    def +=(t : (GeneratableGraph, ToggleButton)) {
      val (graph, button) = t
      contents += button
      if (toggled == null) setToggled(button)
    }
  }
  
  val sidebar = new BoxPanel(Orientation.Vertical) {
    preferredSize = new Dimension(150, 10)
    
    val dimension = new Dimension(1000,30)
    
    genGraphs.zipWithIndex.foreach { case (graph, num) =>
      graphButtons += graph -> new ToggleButton() {
        val b = this
        action = Action("Graph "+(num+1)) {
          graphButtons.setToggled(b)
          graphpanel.setModel(graph.model)
          graphpanel.repaint()
        }
        maximumSize = dimension
        focusable = false
        
      }
    }
      
    contents += new ScrollPane(graphButtons) {
      border = Swing.EmptyBorder(0)
    }
    
    
  }
  
  
  contents = new BorderPanel() {
    layout(sidebar) = BorderPanel.Position.West
    layout(Component.wrap(graphpanel)) = BorderPanel.Position.Center
  }
  
  size = new Dimension(800,600)
  
}

class GeneratableGraph(path : Derivation.Path[HypergraphProduction]) {
  
  lazy val model : DefaultVisualizationModel[Vertex, Hyperedge] = makeModel()
  
  def makeModel() = {
    val graph = HypergraphGenerator(new OrderedHypergraph(), path)
    val pseudoGraph = graph.asInstanceOf[Graph[Vertex, Hyperedge]]
    
    object glayout extends StaticLayout[Vertex, Hyperedge](pseudoGraph, 
        new RandomLocationTransformer(new Dimension(500, 500)), new Dimension(500, 500))
      with AverageEdgeLayout[Vertex, Hyperedge]
    
    println("Generated")
    new DefaultVisualizationModel(glayout)
  }
}